src/ports/postgres/modules/graph/bfs.py_in

```diff
@@ -0,0 +1,498 @@
+                )
+            ) {count_qry_undirected_init}
+        """.format(**locals()))[0]['count']
+        
+        # Main loop for traversing the graph
+        while vct > 0 and curr_dist_val < max_distance:
+
+            # The loop consists of two steps:
+            # 1) Disover and store all nodes that are linked to nodes found in 
+            #    the immediate previous iteration of the loop that have not already 
+            #    been found in all previous iterations  
+            # 2) Check for any nodes linked to those discovered in Step 1 above
+            #    that have not yet been discovered
+            # 
+            # If a node has multiple possible parents then the parent with the 
+            # smallest ID is chosen for output
+
+            # In the directed graph case only nodes in the {dest} column of 
+            # the edge table are searched to find new nodes reachable from 
+            # previously discovered nodes
+
+            # In the undirected graph case edges are treated as non-directional 
+            # (or bidirectional). Nodes in both the {src} and {dest} columns of 
+            # the edge table are searched to find new nodes reachable from 
+            # previously discovered nodes.
+            # This approach does NOT require the user to provide a forward edge
+            # and a reverse edge between the same two nodes to indicate the 
+            # graph's undirected nature. However, it will work in that scenario
+            # as well.
+
+            insert_qry_undirected_part = ""
+            count_qry_undirected_part = ""
+            
+            if not directed:
+                insert_qry_undirected_part = """ UNION
+                    SELECT {grp_comma} 
+                        {src} AS {vertex_id}, 
+                        {curr_dist_val}+1 AS {dist_col}, 
+                        {dest} AS {parent_col}
+                    FROM {edge_table} 
+                    WHERE (
+                        ({grp_comma} {dest}) IN (
+                            SELECT {grp_comma} {vertex_id} FROM {out_table} 
+                            WHERE {dist_col}={curr_dist_val}
+                        ) 
+                        AND 
+                        ({grp_comma} {src}) NOT IN (
+                            SELECT {grp_comma} {vertex_id} FROM {out_table}
+                        )
+                    )
+                """.format(**locals())
+                
+                count_qry_undirected_part = """ OR (
+                    ({grp_comma} {dest}) IN ( 
+                        SELECT {grp_comma} {vertex_id} FROM {out_table} 
+                        WHERE {dist_col}={curr_dist_val}
+                    ) 
+                    AND 
+                    ({grp_comma} {src}) NOT IN (
+                        SELECT {grp_comma} {vertex_id} FROM {out_table}
+                    )
+                )
+                """.format(**locals())
+
```

Comment:
The `if not directed:` code block should be pulled out of the while loop. I think the only change to these strings when moved out is: `{curr_dist_val}` -> `{{curr_dist_val}}` (even if this doesn't work, there should be a simple way to handle it).